vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/dataview/CheckboxGroupListDataViewImpl.java

```diff
@@ -0,0 +1,152 @@
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class CheckboxGroupListDataViewImpl<T> implements CheckboxGroupListDataView<T> {
+
+    private int itemsSize = 0;
+    private Set<SizeChangeListener> sizeChangeListeners;
+    private CheckboxGroup<T> component;
+
+    public CheckboxGroupListDataViewImpl(CheckboxGroup<T> checkboxGroup) {
+        this.component = checkboxGroup;
+    }
+
+    @Override
+    public T getNextItem(T item) {
+        return getNextOrPreviousItem(item, itemsList -> {
+            int itemIndex = itemsList.indexOf(item);
+            if (itemIndex != -1 && itemIndex < itemsList.size() - 1) {
+                return itemsList.get(itemIndex + 1);
+            }
+            return null;
+        });
+    }
+
+    @Override
+    public boolean hasNextItem(T item) {
+        return getNextItem(item) != null;
+    }
+
+    @Override
+    public T getPreviousItem(T item) {
+        return getNextOrPreviousItem(item, itemsList -> {
+            int itemIndex = itemsList.indexOf(item);
+            if (itemIndex > 0) {
+                return itemsList.get(itemIndex - 1);
+            }
+            return null;
+        });
+    }
+
+    @Override
+    public boolean hasPreviousItem(T item) {
+        return getPreviousItem(item) != null;
+    }
+
+    @Override
+    public CheckboxGroupListDataView<T> withFilter(SerializablePredicate<T> filter) {
+        CheckboxGroupListDataView<T> dataView = withFilterOrOrder(
+                dataProvider -> dataProvider.setFilter(filter));
+        // fire size change event if size has changed
+        getDataSize();
+        return dataView;
+    }
+
+    @Override
+    public CheckboxGroupListDataView<T> withSortComparator(SerializableComparator<T> sortComparator) {
+        return withFilterOrOrder(dataProvider -> dataProvider.setSortComparator(sortComparator));
+    }
+
+    @Override
+    public Stream<T> getAllItems() {
+        DataProvider<T, ?> dataProvider = getDataProvider();
+        return dataProvider.fetch(new Query<>());
+    }
+
+    @Override
+    public int getDataSize() {
+        int size = getDataProvider().size(new Query<>());
+        if (size != itemsSize && sizeChangeListeners != null) {
+            sizeChangeListeners.forEach(listener -> listener.sizeChanged(
+                    new SizeChangeEvent<>(this.component, size)));
+            itemsSize = size;
+        }
+        return size;
+    }
+
+    @Override
+    public boolean dataContainsItem(T item) {
+        // InMemoryDataProviders do not override getId(), so rely on equals() method
+        return getAllItems().collect(Collectors.toList()).contains(item);
+    }
+
+    @Override
+    public T getItemOnRow(int row) {
+        if (row < 0) {
+            throw new IndexOutOfBoundsException("Row number should be zero on greater");
+        }
+
+        List<T> itemsList = getAllItems().collect(Collectors.toList());
```

Comment:
Use `getDataProvider().getItems()` as Collection can be used before return where we can do `new ArrayList(items).get(row)`

Reply:
Done.